compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -2532,6 +2547,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         }
                     }
                 }
+                // We special case the situation where `Pin::new` wouldn't work, and isntead
+                // suggest using the `pin!()` macro instead.
+                if let Some(new_rcvr_t) = Ty::new_lang_item(self.tcx, *rcvr_ty, LangItem::Pin)
+                    && !alt_rcvr_sugg
+                    && !unpin
+                    && sym::as_ref != item_name.name
+                    && *pin_call != ""
```

Comment:
Could you add some comments for these conditions justifying them? 

Saves future readers having to work out why some of these checks are necessary for the suggestion - e.g.
"check the item isn't `as_ref`, it wouldn't make sense with that because..."